app/controllers/sys_controller.rb

```diff
@@ -83,19 +83,41 @@ def fetch_changesets
   end
 
   def repo_auth
-    @project = Project.find_by_identifier(params[:repository])
-
-    if (%w(GET PROPFIND REPORT OPTIONS).include?(params[:method]) &&
-        @authenticated_user.allowed_to?(:browse_repository, @project)) ||
-       @authenticated_user.allowed_to?(:commit_access, @project)
+    project = Project.find_by_identifier(params[:repository])
+    if authorized?(project, @authenticated_user)
       render text: 'Access granted'
-      return
+    else
+      render text: 'Not allowed', status: 403 # default to deny
+    end
+  end
+
+  private
+
+  def authorized?(project, user)
+    if git_auth?
+      authorized_with_git?(project, user, params[:uri], params[:location])
+    else
+      authorized_with_subversion?(project, user, params[:method])
     end
+  end
```

Comment:
Not sure about the scope here, but this will limit the usability of the repo_auth method to git and subversion. Other's will have problems adapting the method.